todo_app/app.py

```diff
@@ -8,4 +8,77 @@
 
 @app.route('/')
 def index():
-    return 'Hello World!'
+    if "todoes" in session:
+        if "removed" in session:
+            return render_template("index.html",todos=session["todoes"],added=session["added"],complete=session["complete"],removed=session["removed"])
+        else:
+            return render_template("index.html",todos=session["todoes"],added=session["added"])
+    else:
+        return render_template("index.html")
+
+@app.route('/add', methods=["POST"])
+def add():
+    if "complete" not in session:
+        session["complete"]=[]
+        complete=[]
+    if "newTask" in request.form:
+        if "todoes" in session:
+            todoes = session["todoes"]
+            todoes.append(request.form["newTask"])
+            session["added"] = [request.form["newTask"]]
+            session["todoes"] = todoes
+        else:
+            session["todoes"] = [request.form["newTask"]]
+            session["added"] = [request.form["newTask"]]
+    if "todoes" not in session:
+        return render_template("index.html")
+    elif "removed" not in session:
+        return render_template("index.html",todos=session["todoes"],added=session["added"],complete=session["complete"])
+    else:
+        return render_template("index.html",todos=session["todoes"],added=session["added"],complete=session["complete"],removed=session["removed"])
+
+
+@app.route('/complete', methods=["POST"])
+def complete():
+    if "todoes" not in session:
+        return render_template("index.html")
+
+    if "complete" not in session:
+        session["complete"]=[]
+        complete=[]
+    else:
+        complete=session["complete"]
+
+    delTaskNums = []
+    for key in request.form:
+       if key.startswith("complete_"):
+           taskNumString = key[9:]
+           try:
+               realTaskNum = int(taskNumString)
+               if realTaskNum >= 0:
+                   if realTaskNum < len(session["todoes"]):
+                       delTaskNums.append(realTaskNum)
+           except ValueError:
+               pass
+    delTaskNums.reverse()
+    todoes = session["todoes"]
+    removed=[]
+    for delTask in delTaskNums:
+        removed.append(todoes[delTask])
+        cell=[todoes.pop(delTask)]
+        cell.append(datetime.now().strftime("%d/%m/%Y %H:%M:%S"))
+        complete.insert(0,cell)
```

Comment:
This is quite complex - definitely worth pulling out into a function! Is there any reason you haven't continued to use the functions set up in session_items? Even if you want to take your own approach as here, it would be nice to store the functionality in another file like that to separate out the routing logic from the session manipulation logic